[ao][quant] adding gpu quant primitives to ao/experimental

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112242
* #112531

Summary: adding the primitives for gpu quantization to ao/experimental,
these api's have no BC guarantee's and are subject to change. Note I to
add the experimental folder to the inline dynamo/inductor file list to
get the tests to pass

Test Plan: python torch/ao/quantization/experimental/_gpu_quantization/test.py

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

HDCharles committed:
[ao][quant] adding gpu quant primitives to ao/experimental

Summary: adding the primitives for gpu quantization to ao/experimental,
these api's have no BC guarantee's and are subject to change. Note I to
add the experimental folder to the inline dynamo/inductor file list to
get the tests to pass

Test Plan: python torch/ao/quantization/experimental/_gpu_quantization/test.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vadimkantorov commented on torch/ao/quantization/experimental/_gpu_quantization/quant_primitives.py:
```diff
@@ -0,0 +1,384 @@
+    n_bits = 8
+    # if the shape of t is [B, N, K], the shape of scales will be [B, N, 1]
+
+    scales = t.abs().max(dim=-1, keepdim=True)[0]
```
should this instead use `amax` directly: `t.abs().amax(dim=-1, keepdim=True)`?

HDCharles replied:
yeah i can change that

cpuhrsch commented on torch/ao/quantization/experimental/_gpu_quantization/test.py:
```diff
@@ -0,0 +1,1024 @@
+# mypy: ignore-errors
```
Please move tests under the toplevel test folder

HDCharles replied:
OK

cpuhrsch commented on torch/ao/quantization/experimental/_gpu_quantization/utils.py:
```diff
@@ -0,0 +1,123 @@
+    return s
+
+
+class IntMMModule(torch.nn.Module):
```
Is this still needed?

HDCharles replied:
no, i'll remove it

cpuhrsch commented on torch/ao/quantization/experimental/_gpu_quantization/smoothquant.py:
```diff
@@ -0,0 +1,455 @@
+#         self.x_running_abs_max = W_absmax
+
+
+# # wrapper of https://github.com/facebookresearch/fairscale/
```
There's a lot of commented-out code here - Should that be removed?

cpuhrsch commented:
Test failures are real and mostly related to docs, API visibility and a dependency on matplotlib

cpuhrsch commented on torch/ao/quantization/experimental/_gpu_quantization/smoothquant.py:
```diff
@@ -0,0 +1,455 @@
+"""
+Testing out accuracy-only implementation of SmoothQuant
```
As far as I know this isn't a fully faithful reimplement of smoothquant? Maybe we should revisit this later.

cpuhrsch commented:
Marking this as WIP and converting it to draft

HDCharles committed:
Update on "[ao][quant] adding gpu quant primitives to ao/experimental"


Summary: adding the primitives for gpu quantization to ao/experimental,
these api's have no BC guarantee's and are subject to change. Note I to
add the experimental folder to the inline dynamo/inductor file list to
get the tests to pass

Test Plan: python torch/ao/quantization/experimental/_gpu_quantization/test.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
